### PR TITLE
feat(ai-engine): Module A.2 — document summarisation (summary, glossary, outline)

### PR DESCRIPTION
**Pull request by chaitanya21kumar**
## What this adds

Module A.2 — the summarisation layer over the parsed document from A.1. Given a
ParsedDocument, it derives learner-facing insights (a summary, key takeaways, a
glossary, and a course outline) via a language model, returned as a portable
DocumentInsights JSON contract.

Two endpoints:
- `POST /summarize` — takes a ParsedDocument (the /ingest output) → DocumentInsights
- `POST /summarize/file` — takes a PDF/PPTX upload, parses + summarises in one call

## Model layer (per ADR-0002)

Inference goes through the OpenAI-compatible chat-completions contract, so the
provider is a config choice: a local Ollama (offline default) or a hosted
provider during development. We develop against a provider serving the same open
models we'd self-host for production (Llama 3.1 8B), so what we validate now
carries over. Issue #7's self-hosted, no-cloud requirement stays the production
goal, swapped in behind the same interface.

## Design notes

- DocumentInsights is kept separate from ParsedDocument (deterministic parse vs
  generative output): the parse contract stays frozen, and a single failed
  section degrades to a warning instead of failing the whole request.
- Source-grounded prompts (no hallucination) with an injection guard; JSON output
  validated with Pydantic.
- Error handling: gateway unreachable / bad key → 503, timeout → 504, empty
  document → 400, rate-limit (429) → retried with capped backoff then 503.

## How it was tested

- 33 unit tests (offline, gateway mocked): happy path, empty doc, unreachable,
  connect/read timeouts, rate-limit retry + exhaustion, single-section
  degradation, list-summary coercion, file upload, unsupported type, flattening.
- Live end-to-end on both a hosted provider (Llama 3.1 8B) and local Ollama; a
  3-page and a 10-page document both summarise cleanly (200, all sections, no
  warnings).

Screenshots below.

Refs #7

### CODE REVIEW

**Comment by chaitanya21kumar** [comment]
Proof — all 33 unit tests passing, and a live /summarize/file run returning the full insights:
pr56-live-summarize
pr56-tests-passing

File: ai-engine/app/summarization/llm_client.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    """Pull the assistant message out of a chat completion and parse it as JSON."""
+    try:
+        body = response.json()
+    except ValueError as exc:  # includes json.JSONDecodeError
+        raise LLMBadResponse(f"Gateway response body was not JSON: {exc}") from exc
+    choices = body.get("choices") if isinstance(body, dict) else None
+    if not choices:
+        raise LLMBadResponse("Gateway response contained no choices.")
+    content = (choices[0].get("message") or {}).get("content")
+    if not content:
```
[the remark this answers is not in the document]
Comment (chaitanya21kumar) [reply ai-engine/app/summarization/llm_client.py:131]: Done — hardened the parsing in `_extract_content` to check that `choices` is a non-empty list and that its first element and `message` are dicts before indexing, so a malformed reply becomes a clean `LLMBadResponse` (degraded to a warning by the pipeline) instead of an AttributeError. Pushed in 84c00b2.

File: ai-engine/app/summarization/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    glossary = await _generate_section(
+        client, config,
+        user=prompts.glossary_prompt(source_text),
+        response_model=_GlossaryResponse, label="glossary", warnings=warnings,
+    )
+    outline = await _generate_section(
+        client, config,
+        user=prompts.outline_prompt(source_text),
+        response_model=_OutlineResponse, label="outline", warnings=warnings,
+    )
```
[the remark this answers is not in the document]
Comment (chaitanya21kumar) [reply ai-engine/app/summarization/pipeline.py:182]: Kept these sequential on purpose. Development runs against a hosted free tier and production self-hosts a single local model; running the three calls concurrently bursts the per-minute rate limit (we already saw 429s on larger documents) and a single local model serialises them anyway, so `gather` would not reduce wall-clock time there. The retry/backoff handles transient limits. Worth revisiting with `asyncio.gather` on a paid or scaled deployment where concurrency clearly wins.

**Comment by siddhishinde0723** [comment]
🤖 **AI Code Review Summary**

### 📝 Overview
The provided code is architecturally sound, following clean separation of concerns by isolating the transport/routing layer from the business logic and LLM orchestration. The use of a provider-agnostic interface via OpenAI-compatible contracts is excellent for long-term maintainability. The main risks involve potential instability in model output validation (JSON schema drift) and the lack of explicit logging for production observability. GLOSSARY = "glossary" OUTLINE = "outline" ``` **Location**: `app/summarization/pipeline.py:34` **Issue**: You are using `field_validator(mode="before")` to coerce types for `summary` and `key_takeaways`. While resilient, this hides potential "model drift" where the model stops following the JSON schema instructions. **Fix**: Add a `Counter` or log warning when coercion occurs so you can monitor if the model is frequently failing to output the requested format. **Location**: `tests/test_pipeline.py` **Issue**: `flatten_document` handles tables and lists, but complex nested structures or empty blocks might produce unexpected whitespace. **Fix**: Add a test case for a document containing only empty pages or empty blocks to ensure `flatten_document` returns an empty string consistently, preventing the `EmptyDocumentError` from being triggered by "junk" data. *   **Excellent work** on extracting the `service.py` ingestion logic. This prevents the "router drift" you mentioned in the docstrings. *   The `httpx.MockTransport` usage in `tests/test_summarize.py` is the gold standard for testing AI-integrated pipelines without requiring real LLM calls. Keep this pattern as you add Module A.3.

### 🚩 Top Findings

#### [MAJOR] Potential race conditions on `app.state`
- **Location**: `app/main.py:41-42`
- **Issue**: The `lifespan` initialization of `http_client` and `llm_client` sets them to `None` and then populates them. However, if a request hits an endpoint (e.g., `/summarize`) before the lifespan startup logic completes, the application will crash or raise an unhandled `AttributeError`.

#### [MAJOR] Lack of observability for model failures
- **Location**: `app/summarization/pipeline.py:126`
- **Issue**: When a generation fails (mapping to `None` in `_generate_section`), the error is added to a `warnings` list, but it is **not logged**. In a production scenario, silent failures of specific sections (glossary/outline) will be impossible to debug without checking individual API responses.

**Comment by chaitanya21kumar** [comment]
Thank you, ma'am — went through each point.

1. **Coercion hiding model drift:** the validators stay (they stop one odd response from losing a whole section), but they now log a warning whenever they actually coerce `summary`/`key_takeaways`, so recurring drift shows up in the logs instead of silently. — `0b522c7`
2. **Observability for failed sections:** fixed — a degraded section is now logged in `_generate_section`, not just added to the response warnings. — `0b522c7`
3. **Empty-document test:** added — empty/image-only pages flattening to an empty string, and `generate_insights` raising `EmptyDocumentError` before any gateway call. — `f29e153`
4. **Race on `app.state`:** this one's already guarded — the clients are pre-set to `None` in `create_app` (`main.py:51-52`), `get_llm_client` returns a clean 503 if it's ever `None` (`router.py:41-43`), and `/ready` guards it too. ASGI also runs the lifespan startup before any request is served, so a request can't reach the endpoint before the client exists. Happy to add more if you'd like.

All 37 tests green and the SonarCloud quality gate is passing. Thanks again for the review.

**Comment by chaitanya21kumar** [comment]
Thanks for the review — all points addressed:

- **Prompt JSON examples (glossary + outline):** replaced the set-like `{"term", "definition"}` / `{"title", "points"}` with proper key/value shapes so the model has a valid target. — `203aef2`
- **`/ready` README line and the package docstring:** reworded to the configurable OpenAI-compatible gateway rather than Ollama specifically. — `fefed22`
- **Fixture leaking the fake client:** the summarisation test fixture now tracks the `AsyncClient`(s) it installs and closes them on teardown. — `cef6f74`

37 tests green.
